src/vmemcache_heap.c

```diff
@@ -198,59 +226,105 @@ vmcache_heap_destroy(struct heap *heap)
+	ASSERTeq(first_header->prev, NULL);
+	ASSERTne(first_header->next, NULL);
+
+	void *second_extent = first_header->next;
+	struct header *second_header = HEADER(second_extent);
+	ASSERTeq(second_header->prev, ext->ptr);
+
+	/* remove the first extent from the list */
+	second_header->prev = NULL;
+
+	/* free the first extent */
+	struct heap_entry he;
+	vmcache_heap_merge(heap, ext, &he);
+	vmcache_insert_heap_entry(heap, &he, &heap->last_extent, IS_FREE);
+
+	__sync_fetch_and_sub(&heap->size_used, ext->size);
+
+	return 0;
+}
+
 /*
  * vmcache_alloc -- allocate memory (take it from the queue)
  *
- * This function returns the number of allocated bytes if successful,
- * otherwise -1 is returned.
+ * It returns the number of allocated bytes if successful, otherwise -1.
+ * The last extent of doubly-linked list of allocated extents is returned
+ * in 'last_extent'.
+ * 'first_extent' has to be zeroed in the beginning of a new allocation
+ * (e.g. when *last_extent == NULL).
  */
 ssize_t
-vmcache_alloc(struct heap *heap, size_t size, struct extent_vec *vec)
+vmcache_alloc(struct heap *heap, size_t size, void **last_extent,
+		struct extent *first_extent)
 {
-	LOG(3, "heap %p size %zu", heap, size);
+	ASSERTne(last_extent, NULL);
+	ASSERT((*last_extent == NULL) ?
+		(first_extent->ptr == NULL && first_extent->size == 0) : 1);
 
-	struct heap_entry he = {NULL, 0};
+	LOG(3, "heap %p size %zu last_extent %p first_extent->ptr %p",
+			heap, size, *last_extent, first_extent->ptr);
+
+	struct heap_entry he, f;
+	size_t extent_size = heap->extent_size;
 	size_t to_allocate = size;
 	size_t allocated = 0;
 
 	util_mutex_lock(&heap->lock);
 
 	do {
-		if (vmcache_heap_get(heap, &he))
+		if (vmcache_get_heap_entry(heap, &he))
 			break;
 
 		size_t alloc_size = roundup(to_allocate + HFER_SIZE,
-						heap->extent_size);
+						extent_size);
 
-		if (he.size >= alloc_size + heap->extent_size) {
-			struct heap_entry f;
+		if (he.size >= alloc_size + extent_size) {
 			f.ptr = VOID_PTR(he.ptr, alloc_size);
 			f.size = he.size - alloc_size;
 
-			vmcache_heap_insert(heap, &f);
+			vmcache_insert_heap_entry(heap,
+					&f, &heap->last_extent, IS_FREE);
 
 			he.size = alloc_size;
 		}
 
-		if (vmcache_heap_vec_insert(vec, &he)) {
+		if (vmcache_insert_heap_entry(heap, &he, last_extent,
+							IS_ALLOCATED)) {
 			util_mutex_unlock(&heap->lock);
 			return -1;
 		}
 
-		/* size without headers */
-		allocated += he.size;
+		if (first_extent->ptr == NULL && to_allocate == size) {
+			first_extent->ptr = *last_extent;
+			first_extent->size = he.size - HFER_SIZE;
+		}
 
-		if (to_allocate <= he.size) {
-			to_allocate = 0;
-			break;
+		/* allocated size without headers */
+		size_t allocated_size = he.size - HFER_SIZE;
+		allocated += allocated_size;
+
+		if (allocated_size > to_allocate &&
+		    allocated_size - to_allocate >= extent_size - HFER_SIZE &&
+		    first_extent->size == extent_size - HFER_SIZE) {
+			vmcache_free_first_extent(heap, first_extent);
```

Comment:
Does it matter that the extent is first? Shouldn't this be "free any extent that has size == extent_size" ?

Reply:
Right, done.